File: backend/app/schemas/messaging.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator


class ConversationCreate(BaseModel):
    other_user_id: int = Field(..., gt=0, description="Target user ID to start conversation with")
    initial_message: Optional[str] = Field(
        None,
        description="Optional initial message text",
    )

    @field_validator("initial_message")
    @classmethod
    def validate_initial_message(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            if not isinstance(v, str):
                raise ValueError("Initial message must be a string")
            v = v.strip()
            if not v:
                raise ValueError("Initial message cannot be empty or whitespace only")
            if len(v) > 5000:
                raise ValueError("Initial message cannot exceed 5000 characters")
        return v


class ParticipantSummary(BaseModel):
    id: int
    email: str
    role: str
    full_name: Optional[str] = None
    company_name: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class MessageCreate(BaseModel):
    body: str = Field(..., description="Message text content")

    @field_validator("body")
    @classmethod
    def validate_body(cls, v: str) -> str:
        if not isinstance(v, str):
            raise ValueError("Message body must be a string")
        v = v.strip()
        if not v:
            raise ValueError("Message body cannot be empty or whitespace only")
        if len(v) > 5000:
            raise ValueError("Message body cannot exceed 5000 characters")
        return v


class MessageResponse(BaseModel):
    id: int
    conversation_id: int
    sender_id: int
    sender_email: Optional[str] = None
    body: str
    is_read: bool
    created_at: datetime
    read_at: Optional[datetime] = None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ConversationResponse(BaseModel):
    id: int
    created_at: datetime
    updated_at: datetime
    other_participant: ParticipantSummary
    last_message: Optional[MessageResponse] = None
    unread_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class ConversationListResponse(BaseModel):
    items: List[ConversationResponse]
    total: int


class MessageListResponse(BaseModel):
    items: List[MessageResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class MarkReadResponse(BaseModel):
    conversation_id: int
    marked_read_count: int

File: frontend/src/types/messaging.ts
/**
 * Messaging TypeScript Definitions
 * Strictly aligned with CareerBridge backend Messaging Pydantic schemas.
 */

export interface ParticipantSummary {
  id: number;
  email: string;
  role: string;
  full_name?: string | null;
  company_name?: string | null;
}

export interface Message {
  id: number;
  conversation_id: number;
  sender_id: number;
  sender_email?: string | null;
  body: string;
  is_read: boolean;
  created_at: string;
  read_at?: string | null;
  updated_at: string;
}

export interface Conversation {
  id: number;
  created_at: string;
  updated_at: string;
  other_participant: ParticipantSummary;
  last_message?: Message | null;
  unread_count: number;
}

export interface ConversationListResponse {
  items: Conversation[];
  total: number;
}

export interface MessageListResponse {
  items: Message[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreateConversationPayload {
  other_user_id: number;
  initial_message?: string;
}

export interface SendMessagePayload {
  body: string;
}

export interface MarkReadResponse {
  conversation_id: number;
  marked_read_count: number;
}

export type WebSocketConnectionState =
  | 'connecting'
  | 'connected'
  | 'disconnected'
  | 'reconnecting'
  | 'error';

export interface WSNewMessageEvent {
  type: 'new_message';
  message: Message;
}

export interface WSMessagesReadEvent {
  type: 'messages_read';
  conversation_id: number;
  reader_id: number;
  marked_read_count: number;
  read_at: string;
}

export interface WSMessageReadEvent {
  type: 'message_read';
  conversation_id: number;
  message_id: number;
  reader_id: number;
  read_at?: string | null;
}

export interface WSPongEvent {
  type: 'pong';
}

export interface WSErrorEvent {
  type: 'error';
  code: string;
  message: string;
}

export type WebSocketEvent =
  | WSNewMessageEvent
  | WSMessagesReadEvent
  | WSMessageReadEvent
  | WSPongEvent
  | WSErrorEvent;
